src/loader/bgw_launcher.c

```diff
@@ -254,13 +276,25 @@ init_database_htab(void)
 
 /* Insert a scheduler entry into the hash table. Correctly set entry values. */
 static DbHashEntry *
-db_hash_entry_create(HTAB *db_htab, Oid db_oid)
+db_hash_entry_create_if_not_exists(HTAB *db_htab, Oid db_oid)
 {
 	DbHashEntry *db_he;
 	bool		found;
 
 	db_he = (DbHashEntry *) hash_search(db_htab, &db_oid, HASH_ENTER, &found);
-	db_he->db_scheduler_handle = NULL;
+	if (!found)
+	{
+		db_he->db_scheduler_handle = NULL;
+		db_he->state = ENABLED;
+		SetInvalidVirtualTransactionId(db_he->vxid);
+		db_he->state_advance_failures = 0;
+
+		/*
+		 * Try to allocate a spot right away to give schedulers priority over
+		 * other bgws
+		 */
+		scheduler_state_trans_enabled_to_allocated(db_he);
```

Comment:
Not saying this is wrong, but it seems like a very implicit way of giving priority. Isn't there a more explicit way of doing this by some kind of ordering or priority scheme? This seems like we are kind of racing to get a spot which seems a bit weird. We could, for instance, alott a number of slots for schedulers, although they could go over that alottment.

I guess it can also be a bit confusing that some states auto-transition, while some don't and there is no single place for transitions.

Reply:
This was mostly done to preserve the old semantics where we had an initial logic that allocated slots for all databases on startup to prevent races between workers started by schedulers and the initial set of schedulers. I'll put a better comment but I think if we want to change this it should be a separate PR.